idaes/apps/grid_integration/bidder.py

```diff
@@ -0,0 +1,850 @@
+#################################################################################
+# The Institute for the Design of Advanced Energy Systems Integrated Platform
+# Framework (IDAES IP) was produced under the DOE Institute for the
+# Design of Advanced Energy Systems (IDAES), and is copyright (c) 2018-2021
+# by the software owners: The Regents of the University of California, through
+# Lawrence Berkeley National Laboratory,  National Technology & Engineering
+# Solutions of Sandia, LLC, Carnegie Mellon University, West Virginia University
+# Research Corporation, et al.  All rights reserved.
+#
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and
+# license information.
+#################################################################################
+import pandas as pd
+import pyomo.environ as pyo
+from pyomo.opt.base.solvers import OptSolver
+import os
+from itertools import combinations
+from abc import ABC, abstractmethod, abstractproperty
+
+
+class AbstractBidder(ABC):
+
+    """
+    The abstract class for all the bidder and self-schedulers.
+    """
+
+    @abstractmethod
+    def update_model(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def compute_bids(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def write_results(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def formulate_bidding_problem(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def record_bids(self, *args, **kwargs):
+        pass
+
+    @abstractproperty
+    def generator(self):
+        return "AbstractGenerator"
+
+    def _check_inputs(self):
+
+        """
+        Check if the inputs to construct the tracker is valid. If not raise errors.
+        """
+
+        self._check_bidding_model_object()
+        self._check_n_scenario()
+        self._check_solver()
+
+    def _check_bidding_model_object(self):
+
+        """
+        Check if tracking model object has the necessary methods and attributes.
+        """
+
+        method_list = ["populate_model", "update_model"]
+        attr_list = ["power_output", "total_cost", "generator", "pmin", "default_bids"]
+        msg = "Tracking model object does not have a "
+
+        for m in method_list:
+            obtained_m = getattr(self.bidding_model_object, m, None)
+            if obtained_m is None:
+                raise AttributeError(
+                    msg
+                    + f"{m}() method. The bidder object needs the users to implement this method in their model object."
+                )
+
+        for attr in attr_list:
+            obtained_attr = getattr(self.bidding_model_object, attr, None)
+            if obtained_attr is None:
+                raise AttributeError(
+                    msg
+                    + f"{attr} property. The bidder object needs the users to specify this property in their model object."
+                )
```

Comment:
Would it be better to have an ABC for the bidding_model_object, e.g. `AbstractBiddingModel`. Then we could just check `isinstance(self.bidding_model_object, AbstractBiddingModel)`. I see this is an issue further in the code -- might best to be just enforced through ABCs.

As-it-is, this code defines the expected interface. This might be better:
```python
class AbstractBiddingModel(ABC):
    @abstractmethod
    def populate_model(self, b):
        '''
        documentation about what the user should implement
        '''
        pass

    @abstractmethod
    def update_model(self, b, implemented_shut_down, implemented_start_up, implemented_power_output):
        '''
        documentation about what should be implemented
        '''
        pass

    @property
    @abstractmethod
    def power_output(self):
        '''
        documentation about this property
        '''
        pass

    @property
    @abstractmethod
    def total_cost(self):
        '''
        documentation about this property
        '''
        pass

    @property
    @abstractmethod
    def generator(self):
        '''
        documentation about this property
        '''
        pass

    @property
    @abstractmethod
    def pmin(self):
        '''
        documentation about this property
        '''
        pass

    @property
    @abstractmethod
    def default_bids(self):
        '''
        documentation about this property
        '''
        pass

```

Reply:
@bknueven Thanks, Ben! This makes sense. It seems to be more changes than other suggestions. Should I open another PR to address this?

Reply:
I don't think this is critical -- if you want to address it in a follow on PR I think that is okay.

Reply:
If you go with a follow-up PR, we should create an issue with a checklist of everything we want in the follow-up PR(s).